src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -174,16 +174,105 @@ private sealed class RunningInstance
+                                new Thread(WatchForFileSystemEventsThreadStart) { IsBackground = true }.Start(new object[] { runLoopStarted, eventStream });
+                                runLoopStarted.Wait();
+
+                                Interlocked.Increment(ref scheduledStreamsCount);
+                                return;
+                            }
+                        }
+                    }
+
+                    // Schedule the EventStream to run on the thread's RunLoop
+                    Interop.EventStream.FSEventStreamScheduleWithRunLoop(eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+                    Interlocked.Increment(ref scheduledStreamsCount);
+                }
+
+                public static void UnscheduleFromRunLoop(SafeEventStreamHandle eventStream)
+                {
+                    if (_watcherRunLoop != IntPtr.Zero)
+                    {
+                        // Always unschedule the RunLoop before cleaning up
+                        Interop.EventStream.FSEventStreamUnscheduleFromRunLoop(eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+                        Interlocked.Decrement(ref scheduledStreamsCount);
+
+                        if (scheduledStreamsCount == 0)
+                        {
+                            // Stop the FS event message pump
+                            Interop.RunLoop.CFRunLoopStop(_watcherRunLoop);
+                        }
+                    }
+                }
+
+                private static void WatchForFileSystemEventsThreadStart(object args)
                 {
-                    if (!_stopping && _watcherRunLoop != IntPtr.Zero)
+                    var inputArgs = (object[])args;
+                    var runLoopStarted = (ManualResetEventSlim)inputArgs[0];
+                    var _eventStream = (SafeEventStreamHandle)inputArgs[1];
+                    // Get this thread's RunLoop
+                    _watcherRunLoop = Interop.RunLoop.CFRunLoopGetCurrent();
+                    Debug.Assert(_watcherRunLoop != IntPtr.Zero);
+
+                    // Retain the RunLoop so that it doesn't get moved or cleaned up before we're done with it.
+                    IntPtr retainResult = Interop.CoreFoundation.CFRetain(_watcherRunLoop);
+                    Debug.Assert(retainResult == _watcherRunLoop, "CFRetain is supposed to return the input value");
+
+
+                    // Schedule the EventStream to run on the thread's RunLoop
+                    Interop.EventStream.FSEventStreamScheduleWithRunLoop(_eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+
+                    runLoopStarted.Set();
+                    try
+                    {
+                        // Start the OS X RunLoop (a blocking call) that will pump file system changes into the callback function
+                        Interop.RunLoop.CFRunLoopRun();
```

Comment:
I suspect this is going to fail the new test I just added in https://github.com/dotnet/corefx/pull/33903.  Today, each watcher gets its own dedicated thread, and creating that thread will flow the current ExecutionContext from the code setting EnableRaisingEvents to true.  Since that thread then raises the callbacks, those callbacks will all be invoked with the ExecutionContext of the parent.  However, now we're multiplexing all events onto a single thread, which means the ExecutionContext will be that of whichever thread causes the run loop to be spun up.  If I'm right, we'll likely need to do some additional work to capture the ExecutionContext for each watcher and then use that context when invoking the callbacks associated with it.

Reply:
This is the failure:
https://mc.dot.net/#/user/MaximLipnin/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/4374be61ba5a09daeeeac9ed2404c01d75f6766d/workItem/System.IO.FileSystem.Watcher.Tests/analysis/xunit/System.IO.Tests.File_Create_Tests~2FFileSystemWatcher_File_Create_MultipleWatchers_ExecutionContextFlowed
```
Unhandled Exception of Type Xunit.Sdk.EqualException
Message :
Assert.Equal() Failure
Expected: 84
Actual:   42
Stack Trace :
   at System.IO.Tests.File_Create_Tests.<FileSystemWatcher_File_Create_MultipleWatchers_ExecutionContextFlowed>b__1_0() in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_true_prtest/src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Create.cs:line 60
   at System.IO.Tests.FileSystemWatcherTest.ExecuteWithRetry(Action action, Int32 maxAttempts) in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_true_prtest/src/System.IO.FileSystem.Watcher/tests/Utility/FileSystemWatcherTest.cs:line 197
   at System.IO.Tests.File_Create_Tests.FileSystemWatcher_File_Create_MultipleWatchers_ExecutionContextFlowed() in /Users/dotnet-bot/j/workspace/dotnet_corefx/master/osx-TGroup_netcoreapp+CGroup_Debug+AGroup_x64+TestOuter_true_prtest/src/System.IO.FileSystem.Watcher/tests/FileSystemWatcher.File.Create.cs:line 34
```

Reply:
It passes now